### PR TITLE
Add support for chunked transfer encoding used for request bodies

### LINKED ISSUE #55: Support streaming request bodies
It'd be nice if there were a way to send a request body without needing to buffer the entire thing in memory first.

### PR DESCRIPTION
**Pull request by adamreichold**
Closes #55

### CODE REVIEW

**Comment by adamreichold** [comment]
@sbstp This is just the infrastructure to write chunked bodies via user code without any library type using it. Would you like me to add any predefined streaming type like JSON?

File: src/request/builder.rs [hunk 1]
```diff
@@ -400,7 +400,7 @@ impl<B: Body> RequestBuilder<B> {
         header_insert(&mut prepped.base_settings.headers, CONNECTION, "close")?;
         prepped.set_compression()?;
         match prepped.body.kind()? {
-            BodyKind::Empty => (),
```
Comment (sbstp) [inline src/request/builder.rs:403]: Should we set the Transfer-Encoding header to chunked?

Comment (adamreichold) [reply src/request/builder.rs:403]: Definitely! Again, forgot this while porting. Thank god for code review...

**Comment by sbstp** [comment]
> @sbstp This is just the infrastructure to write chunked bodies via user code without any library type using it. Would you like me to add any predefined streaming type like JSON?

I think we can wait a bit before doing this. We will need to add some sort of buffering for serializers since most of them write one byte at a time to the stream, we certainly should not be sending chunks of 1 byte.

File: src/request/builder.rs (hunk 1 above)
Comment (adamreichold) [reply src/request/builder.rs]: Fixed.

**Comment by adamreichold** [comment]
> We will need to add some sort of buffering for serializers since most of them write one byte at a time to the stream, we certainly should not be sending chunks of 1 byte.

In zeptohttpc, I just wired in a `BufWriter` between the serializer and the chunked writer to fix this, e.g. https://github.com/adamreichold/zeptohttpc/blob/2b4bf4507d6e06f439e48aa04c7d920d17fe4254/src/body_writer.rs#L113 (I don't this is a limitation even with the default buffer size of 8 kB as streaming only makes sense for larger bodies anyway.)

**Comment by sbstp** [comment]
Simple enough. I would still prefer to do it in a separate pull request though.

**Comment by sbstp** [comment]
LGTM

**Comment by sbstp** [comment]
@adamreichold I had an idea. Perhaps we could offer a default implementation of the `Body::kind` method that defaults to chunked? That way you don't always have to figure out how to implement it.

**Comment by adamreichold** [comment]
> I had an idea. Perhaps we could offer a default implementation of the `Body::kind` method that defaults to chunked? That way you don't always have to figure out how to implement it.

I am skeptical as this not a good default as the user needs to understand what chunking means to e.g. not fall into the one-character-at-a-time trap of typical serializers. I think providing a sensible set of predefined implementations might be a better usability improvement.

**Comment by sbstp** [comment]
Yeah that's true, good point. I'm planning to release all these changes soon and we can improve the JSON and urlencoded bodies later, if that's ok with you.

**Comment by adamreichold** [comment]
> Yeah that's true, good point. I'm planning to release all these changes soon and we can improve the JSON and urlencoded bodies later, if that's ok with you.

I don't think we can do much about URL-encoded forms as the only supported target is `String` but not `io::Write`, i.e. https://docs.rs/url/2.0.0/url/form_urlencoded/trait.Target.html

I will prepare one more MR for streaming large JSON documents, this rounds out the feature set and would make a good release candidate IMHO.

**Pull request merged**